bulk/management/commands/data_quality.py

```diff
@@ -0,0 +1,343 @@
+        total_votes_no_sources = bills.filter(from_organization=chamber, votes__sources=None).count()
+        no_sources_data[chamber_name].append({
+            "chamber": chamber_name,
+            "total_bills_no_sources": total_bills_no_sources,
+            "total_votes_no_sources": total_votes_no_sources}
+        )
+    return no_sources_data
+
+
+def bill_subjects(bills, chambers):
+    bill_subjects_data = defaultdict(list)
+    for chamber in chambers:
+        chamber_name = chamber.name.lower()
+        overall_number_of_subjects = bills.distinct("subject").values_list("subject", flat=True).count()
+        number_of_subjects = bills.filter(from_organization=chamber).distinct("subject").values_list("subject", flat=True).count()
+        number_of_bills_without_subjects = bills.filter(from_organization=chamber, subject=None).count()
+        bill_subjects_data[chamber_name].append({
+            "chamber": chamber_name,
+            "overall_number_of_subjects": overall_number_of_subjects,
+            "number_of_subjects": number_of_subjects,
+            "number_of_bills_without_subjects": number_of_bills_without_subjects}
+        )
+    return bill_subjects_data
+
+def bills_versions(bills, chambers):
+    bill_version_data = defaultdict(list)
+    for chamber in chambers:
+        chamber_name = chamber.name.lower()
+        bills_without_versions = bills.filter(from_organization=chamber, versions=None).count()
+        bill_version_data[chamber_name].append({
+            "chamber": chamber_name,
+            "total_bills_without_versions": bills_without_versions}
+        )
+
+    return bill_version_data
+
+def vote_data(bills, chambers):
+    bill_vote_data = defaultdict(list)
+    for chamber in chambers:
+        chamber_name = chamber.name.lower()
+        # votes without any voters
+        total_votes_without_voters = bills.filter(from_organization=chamber, votes__votes=None).values_list("votes").count()
+        # votes (which do have voters, as to not include above category)
+        #   but where yes/no count do not match actual voters
+        total_votes_where_votes_dont_match_voters = 0
+        bills_with_votes_with_voters = bills.filter(from_organization=chamber).exclude(votes__votes=None)
+        for b in bills_with_votes_with_voters:
+            for vote_object in b.votes.all():
+                total_yes = 0
+                total_no = 0
+                total_absent = 0
+                total_excused = 0
+                total_not_voting = 0
+                total_other = 0
+
+                voter_count_yes = 0
+                voter_count_no = 0
+                voter_count_absent = 0
+                voter_count_excused = 0
+                voter_not_voting = 0
+                voter_count_other = 0
+
+                vote_counts = vote_object.counts.all()
+                votes = vote_object.votes.all()
+                # Parsing through vote_counts
+                for count in vote_counts:
```

Comment:
i'm torn, as i'm worried about the maintainability of this and thinking that yes/no will cover the important cases. 

let's discuss the thoughts / findings you've seen

also, is this very slow?  it seems like it is doing an extra several queries per bill  instead of a prefetch, how does it look in practice?

Reply:
nevermind the prefetch comment, missed that